smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/middlewares/OperationInputBodyMiddleware.kt

```diff
@@ -27,9 +30,15 @@ class OperationInputBodyMiddleware(
         operationStackName: String,
     ) {
         val inputShapeName = MiddlewareShapeUtils.inputSymbol(symbolProvider, model, op).name
+        val outputShapeName = MiddlewareShapeUtils.outputSymbol(symbolProvider, model, op).name
+        val errorShapeName = MiddlewareShapeUtils.outputErrorSymbolName(op)
         val hasHttpBody = MiddlewareShapeUtils.hasHttpBody(model, op)
         if (hasHttpBody || shouldRender) {
-            writer.write("$operationStackName.${middlewareStep.stringValue()}.intercept(position: ${position.stringValue()}, middleware: ${inputShapeName}BodyMiddleware())")
+            val requestBindings = httpBindingResolver.requestBindings(op)
+            val alwaysSendBodyParam = if (shouldRender) "alwaysSendBody: true" else ""
+            val middlewareTypeName = if (requestBindings.firstOrNull { it.location == HttpBinding.Location.PAYLOAD } != null) "${inputShapeName}BodyMiddleware()" else "ClientRuntime.SerializableBodyMiddleware<$inputShapeName, $outputShapeName, $errorShapeName>($alwaysSendBodyParam)"
```

Comment:
The way this is written makes it difficult to write a unit test and makes it more difficult to reason about what bodyMIddleware to render.  One idea is that we could have a class which determines which BodyMiddleware is needed based on the request binding and returns a type.  That type can then be used as an input to another class which knows how to render the intercept call.  Maybe something like this, you may have another idea on how to break this up?

Reply:
fixed

Reply:
most recent version resolve this!